[export] turn on hybrid symints by default

pianpwk opened the pull request:
Sets `prefer_deferred_runtime_asserts_over_guards=True` for export, so any guards emitted from `SymNode.expect_true` (for example, guards that are implicitly required to be true for an op to succeed) won't lead to constraint violations. Instead these should appear in the graph as runtime asserts, or potentially as replacement expressions for placeholder shapes.

For example, this reshape op should emit s0 * s1 = s2, deferred as a runtime assert.
```
x = torch.randn(4, 8)  # [s0, s1]
y = torch.randn(32)  # [s2]
out = x.reshape(-1) + y
# this emits Eq(s0 * s1, s2), and we represent y's shape as [s0*s1] in the graph.
```

However, other complex guards can still cause export to fail, for instance guards emitted from `SymNode.guard_bool/guard_size_oblivious` (e.g. explicit if-else conditions in user code or lower-level op implementations hit during tracing) can still raise constraint violations. These can be deferred with `allow_complex_guards_as_runtime_asserts=True`. We don't yet make this default, because while this makes export more likely to succeed, it results in non-trivial asserts being emitted that often represent specialization to a variant of the op, or checks related to 0/1 specialization.

We also remove forced specializations for export and kill the `_disable_forced_specializations` flag - now any guard we can't express with Dims/DerivedDims either are handled with Hybrid SymInts, or should be resolved with rewriting or deferring.

Follow up:
Currently, `ShapeEnv._set_replacement()` is called for complex equality expressions (e.g. s2 -> s0*s1 in the example above), and the ExportedProgram stores `s0*s1` in the input placeholder. This isn't checked for validity when the program is run, so an option is to avoid replacement and/or runtime assert on equality.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

pianpwk commented on test/export/test_export.py:
```diff
@@ -1904,17 +1904,6 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         dim0_x = torch.export.Dim("dim0_x", min=3)
         dim1_x = torch.export.Dim("dim1_x", max=8000)
         dynamic_shapes = {"x": (dim0_x, dim1_x)}
-        with self.assertRaisesRegex(
```
we no longer do any specializations

avikchaudhuri replied:
Nice. Might be good to name the exported program and then do some test to ensure what runtime assert was added?

pianpwk replied:
Ah, I should make a note here, this test case does require the complex guards flag to be set, it encounters a guard `Ne(Mod(s0, s0-1), 0)` (special case of reshape) when tracing.

For the cases I've removed, some require this flag, some export fine with just hybrid symints. For those that are fine, the asserts aren't perfect because replacement expressions are set for placeholders, but not checked at runtime. I was planning to leave that as the follow up task (undo replacement, set as runtime assert)

pianpwk commented on torch/_export/serde/serialize.py:
```diff
@@ -1528,6 +1526,15 @@ def deserialize_sym_int(self, s: SymInt) -> Union[int, torch.SymInt]:
+                ):
+                    self.deserialize_sym_int(SymInt.create(as_expr=SymExpr(str(sym))))
+            # then we sympify again using locals to correctly reify with the constructed symbols
             expr = sympy.sympify(val.expr_str, locals=self.symbol_name_to_symbol)
```
This is needed for SymBool expressions like `Eq(u0, 1)`, where u0 isn't bound to any previous node, and directly calling `create_symboolnode()` errors because we haven't tracked range info on u0, so we explicitly deserialize each symbol beforehand.

See `test_sym_bool` in test/export/test_serialize.py for example.

pianpwk commented on torch/export/_trace.py:
```diff
@@ -96,6 +96,7 @@ class ExportDynamoConfig:
         default_factory=set
     )
+    do_not_emit_runtime_asserts = True  # emit runtime asserts after AOTAutograd instead
 
```
This isn't really necessary, and after https://github.com/pytorch/pytorch/pull/128599 isn't that much more efficient, but if we want to reason more about what guards/runtime asserts to emit, I guess this makes it a bit cleaner to do from the export side. Also I guess there's no point in running this twice.

avikchaudhuri replied:
I think comments like this are better added to the code itself.

pianpwk commented on torch/export/_trace.py:
```diff
@@ -625,16 +627,22 @@ def _compiling_state_context():
-    from torch._dynamo import config as _dynamo_config
     from torch._functorch._aot_autograd.input_output_analysis import _graph_output_names
     from torch._guards import detect_fake_mode
 
```
We run the export solver before the runtime asserts pass. Right now doesn't mean much - the export solver is only there for suggested fixes - we won't even get to constraint solving if that's needed. But if in future we want to control what runtime asserts are emitted for export, or use `produce_guards()` to do some simplification on runtime asserts, I think this makes sense.

Also matches how dynamo currently does it.

avikchaudhuri replied:
Same as above, maybe move such comments to code.

pianpwk commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2660,14 +2624,6 @@ def specialize_zero_one(self):
-    def allow_complex_guards_as_runtime_asserts(self):
-        return self.settings.allow_complex_guards_as_runtime_asserts
-
     def check_equal(self, other: "ShapeEnv") -> None:
```
Storing these flags in ShapeEnvSettings becomes a problem with AOTInductor, where the ShapeEnv is reused from exporting with Hybrid Symints = True, but later is frozen for runtime asserts for Inductor. We can't overwrite the settings, and this fails when we end up deferring another guard we want to keep.

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -6277,6 +6213,39 @@ def forward(self, x, y):
+        ep.module()(torch.ones(8), torch.ones(5))
+        with self.assertRaisesRegex(
+            RuntimeError,
+            r"Runtime assertion failed for expression \(s0//2\) \<\= s1",
```
Cool! As a TODO, maybe we should try renaming these symbols to input shapes in the exported program (i.e., build "sources" from the placeholder names and their symbolic shapes). So here we'd get something like `x.shape[0] // 2 <= x.shape[1]`.

Note that I'm not suggesting reusing the shape env sources, but rather build one from (your now nicely named) placeholder names.

pianpwk replied:
Hmm do you mean just prettifying the runtime assert messages, or at some level renaming the symbols? I get the first option, but maybe I'm missing something smarter.

This could be a good BE task for next week?

avikchaudhuri commented on torch/_export/non_strict_utils.py:
```diff
@@ -166,7 +166,7 @@ def make_fake_inputs(
                 tracked_fakes=[],
                 co_fields=co_fields,
-                prefer_deferred_runtime_asserts_over_guards=allow_complex_guards_as_runtime_asserts,
+                prefer_deferred_runtime_asserts_over_guards=True,
```
why keep this flag any more?

pianpwk replied:
The TLDR I think is that this flag helps avoid additional failures, but the asserts emitted are too messy to be on by default.

I guess this can be a longer discussion, but some guards still fail export unless this is specified; hybrid symints only handle the guards going through `SymNode.expect_true`. Guards that go through `guard_bool/guard_size_oblivious` won't get deferred, and these are typically complex guards from tracing lower-level op implementations. An example is in `test_export.py test_reshape_view_helper`:

```
class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x):
        return x.view(x.size(1), -1)
        # torch/_refs/__init__/_reshape_view_helper() will generate guards on reshape kernel(?)
        # Ne(s0, 20), so that reshape isn't no-op
        # Ne(Mod(s0, 20), 0), so that reshape needs to first flatten [s0, 20, 16] -> [s0*20, 16]
        # then split_dim -> [20, s0, 16]

model = Model()
x = torch.rand(1024, 20, 16)
dynamic_shapes = {"x": {0: Dim("batch")}}
```

If that flag is set, then export succeeds and these guards show up in the graph, but I think they seem too specialized to be on by default. There's some test cases that emit a ton more asserts with this flag, and I feel this is similar to exporting with real tensor prop, where we likely succeed but the asserts are a bit of a mess.

avikchaudhuri commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1772,34 +1772,6 @@ def _raise_inconsistencies(self):
             self._inconsistencies.clear()
             raise ValueError(f"The following inconsistencies were found:\n{msg}")
 
-    def _force_specialization(self, s):
```
RIP.

avikchaudhuri reviewed (approved):
Looks great, there seems to be 1 test failure remaining that's repeated several times.

Some comments inline...

pianpwk committed:
[export] turn on hybrid symints by default (#130775)

Summary:
Sets `prefer_deferred_runtime_asserts_over_guards=True` for export, so any guards emitted from `SymNode.expect_true` (for example, guards that are implicitly required to be true for an op to succeed) won't lead to constraint violations. Instead these should appear in the graph as runtime asserts, or potentially as replacement expressions for placeholder shapes.

For example, this reshape op should emit s0 * s1 = s2, deferred as a runtime assert.
```
x = torch.randn(4, 8)  # [s0, s1]
y = torch.randn(32)  # [s2]
out = x.reshape(-1) + y
# this emits Eq(s0 * s1, s2), and we represent y's shape as [s0*s1] in the graph.
```

However, other complex guards can still cause export to fail, for instance guards emitted from `SymNode.guard_bool/guard_size_oblivious` (e.g. explicit if-else conditions in user code or lower-level op implementations hit during tracing) can still raise constraint violations. These can be deferred with `allow_complex_guards_as_runtime_asserts=True`. We don't yet make this default, because while this makes export more likely to succeed, it results in non-trivial asserts being emitted that often represent specialization to a variant of the op, or checks related to 0/1 specialization.

We also remove forced specializations for export and kill the `_disable_forced_specializations` flag - now any guard we can't express with Dims/DerivedDims either are handled with Hybrid SymInts, or should be resolved with rewriting or deferring.

Follow up:
Currently, `ShapeEnv._set_replacement()` is called for complex equality expressions (e.g. s2 -> s0*s1 in the example above), and the ExportedProgram stores `s0*s1` in the input placeholder. This isn't checked for validity when the program is run, so an option is to avoid replacement and/or runtime assert on equality.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

Pull Request resolved: https://github.com/pytorch/pytorch/pull/130775

Reviewed By: avikchaudhuri

Differential Revision: D59778573

Pulled By: pianpwk

clee2000 commented:
@pytorchbot merge -f "merged internally, job failures seem unrelated, bc breakage is intentional"
